### PR TITLE
Add first draft of balenaOS masterclass

### PR DESCRIPTION
**Pull request by ZubairLK**
Signed-off-by: Zubair Lutfullah Kakakhel <zubair@balena.io>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,63 +1,590 @@
...
 
-# 1. Topic 1 Heading
+*Unless explicitly stated, it means accessing the Host balenaOS and not a container*
 
-This is the main exercise section for a topic.
+# Filesystem and partition layout
 
-## 1.1 Subtopic 1.1 Heading
+balenaOS uses a specific filesystem layout. There are 6 partitions:
+- `resin-boot`:Contains the boot files
```
Comment (xginn8) [inline README.md:40]:
```suggestion
- `resin-boot`: Contains the boot files
```

File: README.md [hunk 2]
```diff
@@ -1,63 +1,590 @@
...
+drwxr-xr-x 7 root root 140 Nov  5 08:06 ..
+lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-boot -> ../../sda1
+lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-data -> ../../sda6
+lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-rootA -> ../../sda2
+lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-rootB -> ../../sda3
+lrwxrwxrwx 1 root root  10 Nov  5 08:06 resin-state -> ../../sda5
+root@123123:~#
+```
+
+### State partition and root overlay
```
Comment (xginn8) [inline README.md:130]: Is there meant to be content here?

File: README.md [hunk 3]
```diff
@@ -1,63 +1,590 @@
...
+- `chronyd.service` : A daemon that manages time in the OS via NTP.
+- `NetworkManager.service` : A daemon that manages network connections
+- `ModemManager.service` : A daemon that manages 3g/4g modems connections
+- `balena.service` Runs the balenaEngine(docker) daemon on the device
+- `resin-supervisor.service` : Runs the balena-supervisor container
+- `openvpn.service`: openVPN daemon to connect with balenaCloud VPN.
+
+Other services that are not as commonly asked about
+- `avahi-daemon.service`: avahi advertises network services on the local network.
+- `plymouth*.service`: daemon that manages the balenaOS logo on the screen when booted. splash screen
```
Comment (xginn8) [inline README.md:156]:
```suggestion
- `plymouth*.service`: daemon that manages the balenaOS logo on the screen when booted (splash screen)
```

File: README.md [hunk 4]
```diff
@@ -1,63 +1,590 @@
...
+- Syntax: `nmcli device wifi connect SSID password 'PASSWORD'`
+- Example: `nmcli device wifi connect AndroidAP_1234 password 'nopassword'`
+
+# config.json
+
+`config.json` is a file on the device in the boot partition that is the source of truth about lots of useful bits of information.
+
+*The real file is `/mnt/boot/config.json`. Not `/resin-boot/config.json`*
+
+# pretty-printing config.json.
```
Comment (xginn8) [inline README.md:301]:
```suggestion
## pretty-printing config.json.
```

File: README.md [hunk 5]
```diff
@@ -1,63 +1,590 @@
...
+Edits to `config.json` should not generally be needed. Changing various options in the balenaCloud dashboard results in the supervisor safely editing `config.json` and restarting the specific service that consumes the specific options. Editing by hand manually is an advanced topic discussed later.
+
+# Editing the core OS files at runtime
+
+balenaOS root filesystem is read-only by default for more robustness. But editing the OS files can be quite useful if you want to add more logging/debugging flags while investigating an issue.
+
+We can switch to read-write mode using the following command.
+- `mount -o remount,rw /`
+
+This remounts the filesystem in readwrite mode. Then you can just edit files.
```
Comment (xginn8) [inline README.md:357]:
```suggestion
This remounts the filesystem in read-write mode. Then you can just edit files.
```

File: README.md [hunk 6]
```diff
@@ -1,63 +1,590 @@
...
+
+# Check kernel messages
+
+Use `dmesg` to see the kernel messages
+
+# Check if a kernel config option is enabled.
+
+A copy of the kernel configuration is always available on a device in `/proc/config.gz`
+
+Here is how you would search it if `CONFIG_SPI` is enabled in a dvice.
```
Comment (xginn8) [inline README.md:425]:
```suggestion
Here is how you would search it if `CONFIG_SPI` is enabled in a device.
```

File: README.md [hunk 7]
```diff
@@ -1,63 +1,590 @@
...
+root@123123:~# zcat /proc/config.gz | grep -i config_spi*
+...
+# CONFIG_SPI is not set
+```
+
+So SPI is not enabled in the kernel for this device
+
+# balenaOS on your laptop
+
+We can use docker to spin up a balenaOS container and run bash to poke around it. Useful for various use-case.s
```
Comment (xginn8) [inline README.md:437]:
```suggestion
We can use docker to spin up a balenaOS container and run bash to poke around. Useful for various use-cases.
```

File: README.md [hunk 8]
```diff
@@ -1,63 +1,590 @@
...
+VARIANT="Development"
+VARIANT_ID="dev"
+META_BALENA_VERSION="2.44.0"
+RESIN_BOARD_REV="8fc90cc"
+META_RESIN_REV="7fed82f"
+SLUG="raspberry-pi"
+bash-4.4#
+```
+
+# Advanced: Editing config.json by hand.
```
Comment (xginn8) [inline README.md:496]: To me this section is something we should be heavily discouraging (instead redirecting to https://github.com/balena-io-playground/configizer or similar), but maybe it's useful for debugging?

File: README.md [hunk 9]
```diff
@@ -1,63 +1,590 @@
...
+
+Other services that are not as commonly asked about
+- `avahi-daemon.service`: avahi advertises network services on the local network.
+- `plymouth*.service`: daemon that manages the balenaOS logo on the screen when booted. splash screen
+- `*getty*.service`: provides a login shell over serial/hdmi. Username: `root`
+
+## Checking state of device and/or services:
+`systemctl` is a cli utility as part of systemd that can be used to check various services.
+
+Some common uses
```
Comment (xginn8) [inline README.md:162]:
```suggestion
Some common uses:
```

File: README.md [hunk 10]
```diff
@@ -1,63 +1,590 @@
...
-of the masterclass, which lead to a logical conclusion.
+Whenever 'terminal' is mentioned, this could mean
+- Accessing the device terminal via the dashboard
+- or accessing via `balena ssh`
 
-# 1. Topic 1 Heading
+*Unless explicitly stated, it means accessing the Host balenaOS and not a container*
 
-This is the main exercise section for a topic.
+# Filesystem and partition layout
```
Comment (xginn8) [inline README.md:37]:
```suggestion
# 1. Filesystem and partition layout
```

just to keep with other classes stylistically, ditto all the other headers beneath.

**Review by xginn8 — APPROVED** [review]
lgtm, it would be interesting to see how all the units fit together/depend on one another too (something like `systemd-analyze plot`)

**Comment by xginn8** [comment]
One other small note: you should add yourself to the CODEOWNERS @ZubairLK

**Comment by hedss** [comment]
Please don't merge until my review is up. I'm still awaiting Github to come back to me with an answer. If they don't today, I'll try doing smaller comment blocks :(

**Review by imrehg — DISMISSED** [review]
To block merge until Heds' comments to submit

**Comment by xginn8** [comment]
@imrehg since "requested changes" need to be specifically undone by their creator, I am happy to flip mine (and you can remove yours), such that we don't get stuck making progress here. Thoughts?

**Comment by hedss** [comment]
I've given up on Github coming back with an answer. I'm going to post screenshots of my review here:

screencapture-github-balena-io-projects-balenaos-masterclass-pull-1-files-2019-11-07-10_05_56
screencapture-github-balena-io-projects-balenaos-masterclass-pull-1-files-2019-11-07-10_05_56-2

**Comment by ZubairLK** [comment]
resolved by #2

**Pull request closed without merging**